### PR TITLE
b_b driver: Add support for custom units

### PR DESCRIPTION
**Pull request by slfritchie**
See also: https://github.com/basho/basho_bench/pull/27

### CODE REVIEW

**Comment by reiddraper** [comment]
@slfritchie I'm getting [these](https://gist.github.com/e55fd6c0a48ea0103aca) errors, trying to run basho_bench. Any ideas?

**Comment by slfritchie** [comment]
Reid, did you have the 1debcd6 commit pulled?  My memory says it's that badmatch that you see is a result of the float/int type problem that that commit fixes.  But I could be full of twisty memories, all alike.

**Comment by reiddraper** [comment]
Not getting the errors anymore, I think I hadn't cleaned my environment somehow. I'm not seeing the new units though, do I need a special b_b config to include them? The summary.png looks the same as before.

**Comment by slfritchie** [comment]
Reid, to change the Y axis label on the b_b graph, you need to run `priv/summary.r -i tests/current` with
an additional argument, e.g. `priv/summary.r -i tests/current --ylabel1stgraph gallons/fortnight`

File: bench/basho_bench_driver_moss.erl [hunk 1]
```diff
@@ -254,12 +269,12 @@ do_delete(Host, Url, Headers) ->
...
-do_get(Host, Url, Headers) ->
+do_get(Host, Url, Headers, ReportFun) ->
     case send_request(Host, Url, Headers, get, <<>>, proxy_opts()) of
-        {ok, "200", _Header, _Body} ->
-            ok;
+        {ok, "200", _Header, Body} ->
+            {ok, ReportFun(byte_size(Body))};
         {ok, "404", _Header, _Body} ->
-            ok;
+            {ok, 0.0};
```
Comment (beerriot) [inline bench/basho_bench_driver_moss.erl:292]: `0.0` breaks the spec of `run/4`, which says `{ok, integer()}`

**Comment by beerriot** [comment]
Just tried myself, and seeing this error:

```
=ERROR REPORT==== 3-Apr-2012::15:55:17 ===
Error in process <0.80.0> with exit value: {{case_clause,{{ok,0.000000e+00},{state,1,[{"s3.amazonaws.com",80}],"test",#Fun<basho_bench_driver_moss.3.22268574>}}},[{basho_bench_worker,worker_next_op,1},{basho_bench_worker,rate_worker_run_loop,2}]}
```

**Comment by slfritchie** [comment]
Bryan, I pushed a type fix for the driver.  For the case_clause error in b_b, were you using the "slf-custom-count-units" branch for basho_bench also?  If you'd used "master", I'd expect that case clause to fail.

**Comment by beerriot** [comment]
@slfritchie Whoops - I had a different `origin` set for `basho_bench` than I thought.  No more `case_clause` error now.

**Comment by beerriot** [comment]
I now appear to be in the same state as Reid.  I've set `{moss_measurement_units, byte_sec}.` in my config, and I see it print

```
INFO: Reporting factor = byte_sec
```

but the throughput graph looks exactly the same, numerically (I've also used `kbyte_sec`, so it's not just that my cluster is crazy slow.)

**Comment by slfritchie** [comment]
Bryan, but I mentioned above ... "make results" and "--ylabel1stgraph gallons/fortnight" ... and ... oh, all right, I yield.

**Comment by slfritchie** [comment]
See https://github.com/basho/basho_bench/commit/2d8eb24ea55a42da730ca334edf0494853714bbe

Is there anything else for review?

**Comment by beerriot** [comment]
I think you misunderstand, Scott.  It's not that the label is wrong, it's that I run with `op_sec` and I see a throughput of "2", then I run with `byte_sec` and I see a throughput of "2", then I run with `kbyte_sec` and I see a throughput of "2".  It seems like the setting just isn't working.

I'm running:
- basho_bench: `be71b2c10e35c47887c0f42aecc89e4c2507940d`
- riak_moss: `7afc58d1085d291b0b24577ddb7738f066764eca`

Config:

``` erlang
%% Sample config file for Riak CS basho_bench benchmarking
{moss_measurement_units, byte_sec}.

%{mode, max}.
{mode, {rate,4}}.
{duration, 1}.
{concurrent, 1}.

%% Replace these with your own paths
{source_dir, "/Users/bryan/work/riak_moss/bench"}.
{code_paths, ["deps/ibrowse", "/Users/bryan/work/riak_moss",
              "/Users/bryan/work/riak_moss/deps/stanchion"]}.
{driver, basho_bench_driver_moss}.

%% Replace this with a user you have created.
%% Instructions to create a user are here:
%% https://github.com/basho/riak_moss/wiki/Creating-a-User
{moss_access_key, "R7GGUSKCTH26HDM4CBT6"}.
{moss_secret_key, "umKmBQvgm3gtlF-4nXVekOdtabqeO9l29nuX9w=="}.
{moss_bucket, "test"}.
{moss_disconnect_frequency, 5}.    % # ops before disconnecting HTTP socket
{moss_raw_ip, "s3.amazonaws.com"}. % DO NOT CHANGE
{moss_raw_port, 80}.               % DO NOT CHANGE
%% Replace these with your HTTP proxy's location (i.e. Riak CS)
{moss_http_proxy_host, "localhost"}.
{moss_http_proxy_port, 8080}.
{moss_request_timeout, 999999000}.

%{key_generator, {int_to_bin, {uniform_int, 500}}}.
{key_generator, {uniform_int, 5}}.
%{value_generator_source_size, 12100105}. %% Must be larger than val gen max size.
%{value_generator, {fixed_bin, 1c2100100}}.
{value_generator, {fixed_bin, 1024}}.

{operations, [{insert, 1}, {update, 1}, {delete, 1}, {get, 1}]}.
```

**Comment by slfritchie** [comment]
Bryan, please try again with:

```
{mode, {rate,10}}.
{duration, 1}.
{concurrent, 4}.
```

I'm seeing something weird with the mbyte_sec reporting in the summary.csv file, but the ops/sec, kbyte/sec, and byte/sec items seem to be working correctly.

The rate I'm seeing is 15-25 ops/sec for 1KB objects.  In latencies.csv I see:

```
elapsed, window, n, min, mean, median, 95th, 99th, 99_9th, max, errors
10, 10, 0.046875, 7920, 28481.1, 12875.0, 32862.5, 569572.5, 569732.3, 569740, 0
20, 10, 0.04296875, 7763, 17611.6, 14583.3, 39775.0, 106855.0, 106985.5, 106813, 0
```

But summary.csv says:

```
elapsed, window, total, successful, failed
10, 10, 46.1708984375, 46.1708984375, 0
20, 10, 62.1923828125, 62.1923828125, 0
```

Something isn't quite right about that 3rd column....

**Comment by slfritchie** [comment]
Alllrrrright, may apologies for creating new bugs.  The 0d9e3ff commit fixes the problem that Bryan was seeing where reporting with bytes_sec units would cause problems but any other reporting units would work without error.  Basho_bench was hiding errors from basho_bench_stats, alas.  Once we use floats for all of basho_bench_driver_moss.erl's custom units, things work fine across the board, I believe.

**Comment by beerriot** [comment]
It works! +1

**Pull request merged**